### DISCUSSION TITLE
Modifying a provider in listen

### DISCUSSION THREAD

**Discussion by derolf** [rrousselGit/riverpod#2658] [Q&A]
See this example:

```
extension ProviderListenableX<T> on ProviderListenable<T> {
  AutoDisposeProvider<T> filter(
    bool Function(T state, T next) filter,
  ) =>
      Provider.autoDispose(
        (ref) {
          ref.listen(
            this,
            (_, next) {
              if (filter(ref.state, next)) {
                ref.state = next;
              }
            },
          );
          return ref.read(this);
        },
      );
}
```

I am getting 

```
The provider AutoDisposeProvider<SearchQuery>#35f8d(Mode.idle()) modified
AutoDisposeProvider<GeoCoordinates?>#7ca84 while building.
'package:riverpod/src/framework/element.dart':
Failed assertion: line 478 pos 11: '_debugCurrentlyBuildingElement == null ||
              _debugCurrentlyBuildingElement == this'

When the exception was thrown, this was the stack:
#2      ProviderElementBase._notifyListeners.<anonymous closure> (package:riverpod/src/framework/element.dart:478:11)
#3      ProviderElementBase._notifyListeners (package:riverpod/src/framework/element.dart:489:8)
#4      ProviderElementBase.setState (package:riverpod/src/framework/element.dart:158:7)
#5      ProviderElement.state= (package:riverpod/src/provider/base.dart:331:32)
#6      ProviderListenableX.filter.<anonymous closure>.<anonymous closure> (package:utils/providers.dart:21:21)
#7      ProviderElementBase._listenElement.<anonymous closure> (package:riverpod/src/framework/element.dart:758:42)
#11     ProviderElementBase._notifyListeners.<anonymous closure> (package:riverpod/src/framework/element.dart:546:24)
```

So far, I was under the impression that modifying a provider inside listen is safe. At least, that's what I am doing at different places. So I am wondering why this one fails here...

**Comment by rrousselGit** [comment]
Do you have a full example? It seems fine at first (although the snippet you gave contains numerous bad practices, that's a different topic).

**Reply by derolf** [reply]
Tell me about the bad practice :-)

**Reply by rrousselGit** [reply]
- Providers should always be top-level final variables. Creating providers inside methods is undesirable
- Providers shouldn't be passed as parameter to other objects/functions. By making an extension, the "extended provider" is passed as parameter, so that's not wanted too.

Basically, all provider usage needs to be statically analyzable.

**Reply by derolf** [reply]
Ah don’t worry. That’s only a helper. The actual result provider is stored static final…

isn’t “select” a similar construct?

**Reply by rrousselGit** [reply]
No. It's not based on a provider and has no associated "element"/"ref". 

It works by implementing `ProviderListenable` itself.

**Reply by rrousselGit** [reply]
Either implement ProviderListenable too. Or maybe make a Ref extension, which makes a bit more sense.

**Comment by SzZawadzki** [comment]
I had a similar problem. The problem occurs when providers were reloading, so to fix that I added if in listen to not change state when reloading. (I don't know if it's a good practice)
```
ref.listen(provider, (_, next) {
      if(next.isReloading){
        return;
      }
      state = state.copyWith(x: next);
});
```
